shiny/render/_render.py

```diff
@@ -208,20 +208,22 @@ def __init__(
         # static type checker that it's synchronous. wrap_async() is smart -- if is
         # passed an async function, it will not change it.
         self._fn: RenderPlotFuncAsync = _utils.wrap_async(fn)
+        print(self.__dict__)
 
     def __call__(self) -> Union[ImgData, None]:
         return _utils.run_coro_sync(self._run())
 
     async def _run(self) -> Union[ImgData, None]:
+
         # Reactively read some information about the plot.
         pixelratio: float = typing.cast(
-            float, self._session.input[".clientdata_pixelratio"]()
+            float, self._session.input._map[".clientdata_pixelratio"]()
```

Comment:
I think it's probably a bit dangerous to access `_map` directly from outside the object, because the `Inputs` object sometimes does special handling before accessing `_map`. For example, here it populates an item if it doesn't already exist, so that it doesn't throw an error:

https://github.com/rstudio/py-shiny/blob/8ae23bfc3f2dbd05f2124dac43df3fe0dd3eefd0/shiny/session/_session.py#L863-L871

Maybe we should have a method like this?

```py
def _get(self, use_ns=True): ...
```

Reply:
Or else, `session.parent` and/or `session.root`, this is how I fixed a similar issue in shinywidgets https://github.com/rstudio/py-shinywidgets/pull/50/files#diff-7f04fb8c4f02a66b160fbbf98fa8df1a0aea7e744a9613b9abf6d90a49674fb4R66-R67